### ISSUE TITLE
Optimization: KV cache in inference — generation recomputes the full 512-window at depth 8 for every token

### ISSUE THREAD

**Issue by MatthewLacerda2** [MatthewLacerda2/TinyRefinementModel#153]
**Lane:** partial-cpu — the wiring and the correctness gate are CPU work; only the speedup *number* wants the card.

## What's wrong

`trm/infer.py:85` generates by re-running the **entire padded window** through the **entire model** for every single token:

```python
for i in range(max_new_tokens):
    logits = get_logits_for_token(model, input_ids, valid_len - 1, ...)
```

`run_model_inference` is called at `depth=MAX_STEPS_LIMIT` (8) on a full `MAX_SEQ_LEN=512` array, and only the row at `valid_len - 1` is kept. Every past token's K/V is recomputed from scratch, at every one of the 7 encoder layers and every one of the 8 refinement passes, once per emitted token. There is no KV cache anywhere in the codebase.

Cost is `O(N · S² · K)` where it should be `O(N · S · K)`. At S=512, K=8 that is roughly two orders of magnitude of wasted inference compute, plus we pay full-window cost even when the prompt is 12 tokens long.

## Why the depth loop does *not* block this

Worth stating explicitly, because it looks like it should. `CausalRefiner` runs the shared block K times over the window, so it's natural to assume the cache is invalidated each pass. It isn't:

- The refiner is **causal within the window** by construction (that is the whole distinction from the inert cross-window hunch, `docs/findings/2026-06-13-cross-window-hunch-inert.md`).
- So `z` at position `i` after pass `k` depends only on positions `≤ i`.
- Appending a token at position `n+1` therefore cannot change `z_k` — and hence K/V — at any position `≤ n`, for any pass `k`.

Every past K/V at every depth pass is cacheable. The only twist is **the cache is K deep**: one K/V set per refinement pass rather than one per layer.

Memory is a non-issue — 7 encoder layers + 8 refine passes = 15 layer-equivalents; at 512×960 f16 that is ~2MB each, **~29MB total**. (Which is also the direct answer to "should we consider MLA": there is no KV pressure here to relieve.)

## Acceptance

- **Correctness (CPU, blocking):** cached and uncached generation produce **bit-identical logits** for the same prompt + seed under `FORCE_F32_COMPUTE`. Goes in `tests/core/`. This is the whole gate — a KV cache is a pure refactor, so anything short of exact match is a bug.
- **Speed (GPU, follow-up):** tokens/sec on a 256-token continuation, cached vs current. Expect ≥10× at S=512; report the actual number, no claim without it.

## While in there

`HUNCH_REFRESH_EVERY` / `new_document` (`trm/infer.py:88`) is reasoner-era plumbing — the hunch is proven inert and the default arch is `refiner`. Check whether it still does anything on the live path; if not, it goes with the cache rewrite.

## Not in scope

Attention-variant changes (GQA on the live path, MLA). Those are `ideas`, they need this cache to exist before their benefit is even measurable, and they change what the model is. This issue is `optimization`: same model, same outputs, less compute.

**Comment by MatthewLacerda2** [comment]
Found while scoping this: `trm/infer.py:124` hardcodes `model = UniversalReasoner(LATENT_DIM, nnx.Rngs(0))`, ignoring `MODEL_ARCH` entirely. `trm/runtime/restore.py:53` already has a correct arch-branching `build_model()` — `infer.py` just never calls it.

The two arches have different param trees, so this means `python -m trm.infer` cannot load a `refiner` checkpoint at all — including the current champion (`docs/registry/run_20260719_042625-refiner-base-v1.md`).

Flagging here rather than filing separately because the KV-cache rewrite touches exactly this construction site. If the reasoner gets tombstoned first the whole branch disappears instead; either way `infer.py` should go through `restore.build_model()` rather than naming an arch itself.

**Comment by MatthewLacerda2** [comment]
Re-read this while surveying the model for cheap compute/VRAM wins (2026-08-19). Three notes.

**1. The analysis holds — confirmed independently.** The causality argument is right, and I re-derived the cache shape: 7 encoder blocks + K refine iterations = 13 slots at depth 6 (15 at depth 8), `[512, 15, 64]` × 2 per slot ≈ **25 MB at f16**. No memory pressure whatsoever.

**2. Two pieces split out as their own issues**, because both land immediately and independently while this is a multi-day rewrite:

- **#206** — the tied LM head projects all 512 positions to use one row (`trm/infer.py:91-93`; `token_idx` is traced, so XLA cannot dead-code it). ~22% of per-token generation compute at depth 6, plus a 103 MB f32 transient per token. Subsumed by this issue once the cache lands, but it is a handful of lines today.
- **#207** — the `refresh` static-jit-arg duplicate compile, i.e. the "While in there" bullet at the bottom of this issue. Measured: 2 XLA compilations over 8 generated tokens with the flag alternating, 0 with it pinned. Drop that bullet from here when #207 merges.

**3. Bucketed padding is the cheap partial stand-in, and belongs in this issue rather than its own.** This issue already names it ("we pay full-window cost even when the prompt is 12 tokens long"), so recording the concrete approach here instead of filing a fourth issue that would just get closed as superseded:

`trm/infer.py:114` pads to `MAX_SEQ_LEN` unconditionally, so a 10-token prompt pays a 512-position forward from the very first emitted token. **`instruments/yardstick/yardstick.py:126-132` already solved exactly this** — it buckets to `(64, 128, 256, 512)` and documents the safety argument (causal attention ⇒ trailing pad cannot change earlier logits, so bucketing is purely a compile/memory knob). Reusing that in `generate_text` costs ~4 extra compilations total and cuts early-token work by up to 32×.

Worth noting for whoever picks this up: this is *not* addressed by the model emitting eot early. Halting reduces how many steps you take; bucketing reduces what each step costs. They are different axes, and the KV cache is the thing that fixes the second one properly — bucketing is the version you can have this afternoon.
